[GeoMechanicsApplication] Reorder DOFs of non-diff order U-Pw elements and conditions (#12061)

avdg81 opened the pull request:
**📝 Description**
This PR reorders the degrees of freedom (DOFs) of the non-diff order elements and conditions, such that they are consistent with the ordering applied to the diff order elements and conditions. In essence, this boils down to first enumerating all displacement DOFs, followed by enumerating the pore water pressure DOFs.

**🆕 Changelog**
Changes in this PR include:
- Added utility functions that extract all degrees of freedom of U-Pw elements given the element's nodes and its dimension. These new functions are now used by the diff order elements (and conditions) as well as the non-diff order ones.
- Removed several duplicated or unused (utility) functions.
- Renamed a few utility functions for assembling matrices to make clear which part of the partitioned matrices they consider.
- Re-implemented the utility functions that assemble matrices and vectors (to avoid lots of raw loops).
- Replaced the use of matrix/vector assemble functions where a dimension of 0 was supplied to force the position of addition. Instead, we now use the matrix/vector addition operators, which is much clearer and cleaner.
- Renamed a few function parameters and template type names to be more accurate.
- Added a header file for common (type) aliases.
- Fixed several (minor) code smells, including the removal of commented out code.
- Extended the utility functions for setting up test models. Also added a few helper functions for carrying out common checks in tests.
- Added documentation which describes how the degrees of freedom of the U-Pw elements have been ordered.

avdg81 committed:
Added a utility function for extracting U-Pw DOFs from nodes

The utility function is currently targeted at 2D non-diff order U-Pw
elements. At present, it returns a vector with the correct size,
containing `nullptr`s only.

avdg81 committed:
The utility function no longer returns a vector of `nullptr`s

Instead, it returns a vector of pointers pointing to the
`DISPLACEMENT_X` degree of freedom of the first node.

avdg81 committed:
Expect returned DOFs contain `DISPLACEMENT_X` in certain entries

For each node, we expect the `DISPLACEMENT_X` DOF to come first.

avdg81 committed:
Expect returned DOFs contain `DISPLACEMENT_Y` in certain entries

For each node, we expect the `DISPLACEMENT_Y` DOF to come second.

avdg81 committed:
Expect returned DOFs contain `WATER_PRESSURE` in back entries

For each node, we expect the `WATER_PRESSURE` DOF to come at the end.

avdg81 committed:
Test that three DOFs are associated with node 1

avdg81 committed:
Test that three DOFs are associated with node 2

avdg81 committed:
Test that three DOFs are associated with node 3

avdg81 committed:
Prefer to use `auto` rather than committing to a particular type

avdg81 committed:
Extracted an overload for retrieving the DOFs of nodes

In this way, we were able to eliminate a raw loop.

avdg81 committed:
Reformatted the unit test file

avdg81 committed:
Moved a type alias to its own header file

avdg81 committed:
Extended the utilty function for creating a test model part

The utility function that creates a model part with a single 2D3N
element now receives an optional vector of nodal variables. These are
added to the model part as well as to the new nodes.

This utility function is now used by one of the unit tests for the DOF
utilities.

Also renamed a local variable in the unit test.

avdg81 committed:
Added a unit test for extracting U-Pw DOFs of a 3D element

The test extracts all U-Pw DOFs of a tetra element. So far, it only
checks the size of the returned vector of DOFs.

Note that the interface of the DOF utility function has been extended:
it now also requires the model dimension. When we are dealing with a 3D
model, an extra displacement DOF is added per node. However, note that I
have added `DISPLACEMENT_X` instead of `DISPLACEMENT_Z`, to make a
future check fail.

avdg81 committed:
Check that none of the returned DOF pointers equals `nullptr`

avdg81 committed:
Expect returned DOFs contain `DISPLACEMENT_X` in certain entries

For each node, we expect the degree of freedom associated with
`DISPLACEMENT_X` to be enumerated first.

avdg81 committed:
Expect returned DOFs contain `DISPLACEMENT_Y` in certain entries

For each node, we expect the degree of freedom associated with
`DISPLACEMENT_Y` to be enumerated second.

avdg81 committed:
Expect returned DOFs contain `DISPLACEMENT_Z` in certain entries

For each node, we expect the degree of freedom associated with
`DISPLACEMENT_Z` to be enumerated third.

avdg81 committed:
Expect returned DOFs contain `WATER_PRESSURE` in back entries

avdg81 committed:
Added a few missing DOFs for testing

avdg81 committed:
Expect that four DOFs are associated with node 1

avdg81 committed:
Expect that four DOFs are associated with node 2

avdg81 committed:
Expect that four DOFs are associated with node 3

avdg81 committed:
Expect that four DOFs are associated with node 4

avdg81 committed:
Extracted a local helper for checking the node ID of a DOF

avdg81 committed:
Extracted a local helper for checking the variable of a DOF

avdg81 committed:
Extracted a local helper for checking that no DOFs are null

avdg81 committed:
Simplified some test code

avdg81 committed:
Extracted a helper function for creating an empty model part

avdg81 committed:
Added a utility function for extracting U-Pw DOFs of diff order elements

So far, we only test whether the number of returned DOFs meets our
expectations.

avdg81 committed:
Check that none of the returned DOF pointers is equal to `nullptr`

To make the test pass, we now fill the result with pointers to the DOF
associated with `DISPLACEMENT_X` of the first (second order) node.

avdg81 committed:
Expect returned DOFs contain `DISPLACEMENT_X` in certain entries

For each second order node, we expect the degree of freedom associated
with `DISPLACEMENT_X` to be enumerated first.

avdg81 committed:
Expect returned DOFs contain `DISPLACEMENT_Y` in certain entries

For each second order node, we expect the degree of freedom associated
with `DISPLACEMENT_Y` to be enumerated second.

avdg81 committed:
Expect returned DOFs contain `WATER_PRESSURE` in back entries

For each first order node, we expect the degree of freedom associated
with `WATER_PRESSURE` to be appended to the result.

avdg81 committed:
Expect that three DOFs are associated with node 1

avdg81 committed:
Expect that three DOFs are associated with nodes 2 and 3

And expect that two DOFs are associated with nodes 4 through 6.

avdg81 committed:
The function that creates a 6-noded 2D U-Pw element is more specific

It has been renamed to make that clear. Also, the function no longer
receives collections of first order variables and second order
variables.

avdg81 committed:
Renamed two local variables

avdg81 committed:
Extracted a helper function for adding variables to a model part

Also removed another helper function that had become trivial due to the
above change.

avdg81 committed:
Moved the declaration of two variables closer to their usage

avdg81 committed:
Extracted a local helper function for adding DOFs to nodes

avdg81 committed:
Extracted a local helper function for creating new nodes

avdg81 committed:
Extended the utility function for extracting U-Pw DOFs

The aim of the extension is to support 3D diff order elements in
addition to 2D ones.

Also added a test utility function that creates a 3D diff order tetra.

So far, the new unit test only checks the number of DOFs that is
returned.

Note that the DOF pointer for the third dimension for now is equal to
`nullptr`.

avdg81 committed:
Check that none of the returned DOF pointers is equal to `nullptr`

To make the test pass, we now fill the DOF pointer for the third
dimension with the first DOF pointer of the result (i.e. some DOFs are
repeated).

avdg81 committed:
Check the DOF entries that are expected to have `DISPLACEMENT_{X,Y}`

For each second order node, the entries related to `DISPLACEMENT_X`
should come first, followed by the entries related to `DISPLACEMENT_Y`.

avdg81 committed:
Check the DOF entries that are expected to have `DISPLACEMENT_Z`

For each second order node, the entries related to `DISPLACEMENT_Z`
should come after the entries related to `DISPLACEMENT_Y`.

avdg81 committed:
Check the node IDs associated with the returned DOFs

avdg81 committed:
Reused a function

avdg81 committed:
Refactored some code

avdg81 committed:
U-Pw diff order elements now use the utility function to get the DOFs

avdg81 committed:
Added missing check in unit test

avdg81 committed:
Eliminated a duplicated utility function

A utility function to assemble the U-block vector was implemented twice:
once for elements and once for conditions. The underlying
implementations were, however, identical. Therefore, the condition one
was eliminated in favor of the element one.

avdg81 committed:
Removed three unused condition utility functions

avdg81 committed:
Reordered the DOFs for U-Pw non-diff order elements/conditions

- Updated the assemble utility functions such that they anticipate the
  new DOF ordering.
- `GeoElementUtilities::AssembleUBlockMatrix` is no longer a function
  template. Also eliminated one of the overloads.
- Removed several horizontal separator lines and reformatted some code.

avdg81 committed:
Extracted a function to add a vector to another one with an offset

avdg81 committed:
Extracted a function to add a matrix to another one with offsets

avdg81 committed:
Eliminated a duplication of `AssembleUBlockMatrix`

avdg81 committed:
Reformatted two function calls

avdg81 committed:
Addressed a few code smells

- Removed a commented out `#include`.
- Replaced `typedef` by `using`.
- Removed an alias that had become redundant.

avdg81 committed:
Renamed two functions

avdg81 committed:
Attempt to fix the broken GCC build

Moved a function template closer to the top of the header file in an
attempt to make it visible to the compiler when it is instantiated and
called.

avdg81 committed:
Renamed a few function parameters and template parameters

To make sure that the naming is consistent across the file.

avdg81 committed:
Eliminated a weird way of adding matrices

avdg81 committed:
Addressed two code smells found by SonarQube

Prefer `std::begin` and `std::end` over member functions.

avdg81 committed:
Also reordered the DOFs for the inertia part (continuum elements only)

- Changed the dimensions of the matrices that are needed to calculate
  the mass matrix for U-Pw continuum elements. Since all displacement
  degrees of freedom have been put together, we need one row/column less
  (which used to be there for the water pressure related degree of
  freedom).
- Generalized the element utility function that assembles (better:
  fills) the density matrix.
- Removed an element utility function that was no longer used.

avdg81 committed:
Also reordered the DOFs for the inertia part (interface elements only)

- Removed the calculation of the shape function matrix that took into
  account the previous DOF ordering.
- Changed the dimensions of the matrices that are needed to calculate
  the mass matrix for U-Pw interface elements. Since all displacement
  degrees of freedom have been put together, we need one row/column less
  (which used to be there for the water pressure related degree of
  freedom).

avdg81 committed:
Removed three unused functions and addressed three code smells

avdg81 committed:
Replaced several misused calls to add a vector to another one

The trick here was to pass a dimension of 0 to achieve the effect of
vector addition. We'd better express that intent directly.

A similar trick was carried out for matrix addition.

avdg81 committed:
Added a brief description of the DOF ordering to the documentation

rfaasse commented on applications/GeoMechanicsApplication/custom_conditions/Pw_normal_flux_condition.cpp:
```diff
@@ -83,9 +83,7 @@ void PwNormalFluxCondition<TDim,TNumNodes>::
-    GeoElementUtilities::
-        AssemblePBlockVector< 0, TNumNodes >(rRightHandSideVector,
-                                                rVariables.PVector);
+    rRightHandSideVector += rVariables.PVector;
```
Way cleaner, good catch!

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_microclimate_flux_condition.cpp:
```diff
@@ -169,14 +169,12 @@ void GeoTMicroClimateFluxCondition<TDim, TNumNodes>::CalculateAndAddRHS(
         array_1d<double, TNumNodes>{prod(temporary_matrix, rRightHandSideFluxes)};
-    GeoElementUtilities::AssemblePBlockVector<0, TNumNodes>(
-        rRightHandSideVector, temporary_vector);
+    rRightHandSideVector += temporary_vector;
```
Do we still need to create the temporary_vector variable or is the following enough:
rRightHandSideVector += prod(temporary_matrix, rRightHandSideFluxes);

The same happens on 176-177

avdg81 replied:
Good catch. Indeed, we can eliminate variable `temporary_vector`. Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.cpp:
```diff
@@ -81,8 +81,7 @@ void GeoTNormalFluxCondition<TDim, TNumNodes>::CalculateRHS(Vector& rRightHandSi
         auto normal_flux_on_DOF = normal_flux_on_integration_point * N * weighting_integration_coefficient;
-        GeoElementUtilities::AssemblePBlockVector<0, TNumNodes>(rRightHandSideVector,
-                                                                normal_flux_on_DOF);
+        rRightHandSideVector += normal_flux_on_DOF;
```
or
rRightHandSideVector += normal_flux_on_integration_point * N * weighting_integration_coefficient;

avdg81 replied:
Done. For clarity, I have enclosed the expression that follows `+=` in parentheses.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/U_Pw_face_load_condition.cpp:
```diff
@@ -15,6 +15,7 @@
 
 // Application includes
 #include "custom_conditions/U_Pw_face_load_condition.hpp"
+#include "custom_utilities/element_utilities.hpp"
```
The functionality used from this is shared between elements and conditions. Should it then still be called element_utilities.

avdg81 replied:
Good point. I had the same question when I was making the changes. I would suggest to discuss this in more detail with @rfaasse. Whatever the outcome is, I would propose to delegate that to a separate PR, since it is beyond the scope of this PR.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/README.md:
```diff
@@ -1,3 +1,15 @@
+$$ u_{x,1}, u_{y,1}, u_{x,2}, u_{y,2}, \dots, u_{x,6}, u_{y,6}, p_1, p_2, p_3 $$
+
+where the directions of the displacements are denoted by a suffix (here either $x$ or $y$), and the node number is also added as a suffix. Note that in this case, there are only three pore water pressure DOFs (due to the diff order nature of the element). For a non-diff order quadratic element the above list of DOFs is appended by $p_4$, $p_5$, and $p_6$.
+
```
Maybe a future thing to put a picture here that shows the UU, UP, PU and PP matrix parts and U and P vector parts.
When we add the system of equations here.

avdg81 replied:
Yes, that is a good suggestion. I will add an issue for it on our product backlog. Also, we may want to have a picture that shows the difference between a diff order element and a non-diff order element.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/condition_utilities.hpp:
```diff
@@ -124,80 +124,6 @@ class ConditionUtilities
         }
     }
 
-    //----------------------------------------------------------------------------------------
```
This cleans up nicely!

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/README.md:
```diff
@@ -1,3 +1,15 @@
+
+For simulating soil behavior, the geomechanics application offers continuum elements that take into account the displacement field ($u$) as well as the pore water pressure field ($p_w$). In fact, the two fields can be coupled.
+
+Depending on your needs, you may opt for either so-called "diff order" elements or "non-diff order" elements. The former implies that the order of the interpolation polynomials that will be used for the pore water pressure field is one less than the order of the interpolation polynomials of the displacement field. The latter, on the other hand, assumes one and the same order for the interpolation ...
```
Nice-to-have suggestion, but it might be nice to have a sentence on when you would use diff order (e.g. performance?)

avdg81 replied:
I'd propose to discuss that with @WPK4FEM.

WPK4FEM replied:
This documentation, we will extend bit by bit. When further documenting the "newly emerging" UPw element, I will mention why. ( The answer lies in compatibility of stresses from displacements and pressures. with u 2nd order and p 1st order interpolation, stress and p are both linear therefore compatible ).

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/README.md:
```diff
@@ -1,3 +1,15 @@
+# U-Pw Continuum Elements
```
Clear bit of documentation, thanks for adding this!

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_diff_order_element.cpp:
```diff
@@ -139,7 +139,7 @@ void UpdatedLagrangianUPwDiffOrderElement::CalculateAndAddGeometricStiffnessMatr
 
     // Distribute stiffness block matrix into the elemental matrix
-    this->AssembleUBlockMatrix(rLeftHandSideMatrix, UMatrix);
+    GeoElementUtilities::AssembleUUBlockMatrix(rLeftHandSideMatrix, UMatrix);
```
I'm happy with the name AssembleUUBlockMatrix. Maybe we should consistently rename UMatrix to UUMatrix etc.

avdg81 replied:
I totally agree with your suggestion. Done.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_interface_element.cpp:
```diff
@@ -195,8 +194,6 @@ void UPwSmallStrainInterfaceElement<TDim, TNumNodes>::CalculateMassMatrix(Matrix
 
         this->CalculateJointWidth(Variables.JointWidth, LocalRelDispVector[TDim - 1], MinimumJointWidth, GPoint);
 
-        InterfaceElementUtilities::CalculateNuElementMatrix(Nut, NContainer, GPoint);
```
It's not completely clear why this function could be deleted. Was it a redundant call and did Variables.Nu already contain the matrix?

avdg81 replied:
The deleted function call calculated the N matrix, which was identical to `Variables.Nu` except for the fact that it also contained a few rows and columns with zeroes where the water pressure DOFs were located. Now that the displacement DOFs and water pressure DOFs have been grouped separately (i.e. they are no longer intertwined), `Nut` had become redundant. Instead, we can simply use `Variables.Nu`.

rfaasse replied:
Clear!

WPK4FEM commented on applications/GeoMechanicsApplication/custom_utilities/dof_utilities.h:
```diff
@@ -15,30 +15,70 @@
+}
+
+template <typename NodeRange1, typename NodeRange2>
+std::vector<Dof<double>*> ExtractUPwDofsFromNodes(const NodeRange2& rSecondOrderNodes,
```
Although I understand the logic of NodeRange2 for rSecondOrderNodes, it is confusing as these are the first argument and the U D.O.F. are the first D.O.F. in the element. NodeRangeDisplacementNodes and NodeRangeWaterPressureNodes may be clearer, but also lengthy.

avdg81 replied:
That is a fair point. I have renamed the template types to `DisplacementNodeRange` and `WaterPressureNodeRange`. The function parameters I have renamed to `rDisplacementNodes` and `rWaterPressureNodes`. What do you think about that?

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/dof_utilities.h:
```diff
@@ -15,30 +15,70 @@
     auto result = std::vector<Dof<double>*>{};
-    std::transform(NodePtrRangeBegin, NodePtrRangeEnd, std::back_inserter(result),
-                   [&rDofVariable](const auto& r_node) { return r_node.pGetDof(rDofVariable); });
+    ExtractDofsFromNodes(NodeRangeBegin, NodeRangeEnd, std::back_inserter(result), rDofVariable);
```
What is the reason for extracting this extra function? I think it might be more tricky this way to separate e.g. the std::back_inserter from the std::transform. When looking at the signature of the called function, I could probably by mistake call it with result.begin() instead of std::back_inserter. 

Of course this might be more of a preference issue, so I leave it up to you.

avdg81 replied:
I have designed the interface of the called function in a similar way as `std::transform`, in that it receives an output iterator. In that way, we can fill any container (vectors, arrays, sets, ...). In this particular case, the interface allowed me to easily append the result to an existing vector. See line 72 of `dof_utilities.h`. Without the overload that receives an output iterator, I would have to create a temporary vector that contains the elements to be appended and then append them to the other vector. That is more complex and less efficient than using an output iterator. Also, having an overload with an output iterator is more in line with the STL algorithms. In my opinion, it is the programmer's responsibility to understand the difference between passing `result.begin()` and `std::back_inserter(result)`.

rfaasse replied:
Clear, thank you!

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/dof_utilities.h:
```diff
@@ -15,30 +15,70 @@
+    for (const auto& r_node : rSecondOrderNodes) {
+        for (const auto& r_variable : displacement_variables) {
+            result.push_back(r_node.pGetDof(r_variable.get()));
+        }
```
We could make this inner loop another transform with back inserter.

avdg81 replied:
Good catch. I have replaced the inner raw loop by a call to `std::transform`.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/dof_utilities.h:
```diff
@@ -15,30 +15,70 @@
+std::vector<Dof<double>*> ExtractUPwDofsFromNodes(const NodeRange& rNodes, std::size_t ModelDimension)
 {
-    return ExtractDofsFromNodes(std::begin(rNodePtrs), std::end(rNodePtrs), rDofVariable);
+    return ExtractUPwDofsFromNodes(rNodes, rNodes, ModelDimension);
```
Nice re-use!

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/dof_utilities.h:
```diff
@@ -15,30 +15,70 @@
+template <typename NodeRange>
+std::vector<Dof<double>*> ExtractDofsFromNodes(const NodeRange& rNodes, const Variable<double>& rDofVariable)
+{
+    return ExtractDofsFromNodes(std::begin(rNodes), std::end(rNodes), rDofVariable);
```
Nice that you added this 'range'-version!

WPK4FEM commented on applications/GeoMechanicsApplication/custom_utilities/element_utilities.hpp:
```diff
@@ -249,132 +223,55 @@ typedef std::size_t IndexType;
+    template <unsigned int TDim, unsigned int TNumNodes>
     static inline void AssembleUPBlockMatrix(Matrix& rLeftHandSideMatrix,
-                                             const BoundedMatrix<double,TDim*TNumNodes,TNumNodes>& UPBlockMatrix)
+                                             const BoundedMatrix<double, TDim * TNumNodes, TNumNodes>& rUPBlockMatrix)
```
For the matrix dimension, I wonder if it is better to refer to a number of displacement D.O.F. and a number of water_pressure D.O.F. With TNumNodes here as second size, I suspect that we need another one for the diff order element.

avdg81 replied:
Here, you've touched upon a much bigger issue. In a nutshell, what I have changed is that the interfaces of the assemble functions are now similar to the one of `AssembleUUBlockMatrix`. That is, the template types have been changed, and all functions have been updated. Since we know where the submatrix/subvector needs to be added, we can either hard-code it (when no offset is required) or calculate the offset(s) from the dimensions of the given matrices/vectors.

Again, this cleans up a lot of code. Thanks for bringing this up.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_utilities/element_utilities.hpp:
```diff
@@ -249,132 +223,55 @@ typedef std::size_t IndexType;
-                                            const BoundedMatrix<double,TNumNodes,TNumNodes> &PBlockMatrix)
+    template <unsigned int TDim, unsigned int TNumNodes>
+    static inline void AssemblePPBlockMatrix(Matrix& rLeftHandSideMatrix,
+                                             const BoundedMatrix<double, TNumNodes, TNumNodes>& rPBlockMatrix)
```
rPPBlockMatrix for consistent naming?

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_dof_utilities.cpp:
```diff
@@ -169,6 +193,110 @@ KRATOS_TEST_CASE_IN_SUITE(VariableTypeAndNodeIDsMustMatchWhenExtractingDofsFromN
+    auto& r_model_part = ModelSetupUtilities::CreateModelPartWithASingle2D6NUPwDiffOrderElement(model);
+
+    const auto second_order_nodes = r_model_part.Elements().front().GetGeometry();
+    const auto first_order_nodes =
```
In the time integration schemes we introduces first and second order variables, because we have Pw and T type variables in the same scheme. The UPw element always has U and Pw D.O.F so maybe refer to u_nodes and pw_nodes.

avdg81 replied:
Renamed the variables to `displacement_nodes` and `water_pressure_nodes`, respectively.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/element_utilities.hpp:
```diff
@@ -52,23 +44,6 @@ typedef std::size_t IndexType;
 
-//----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-    template< unsigned int TDim, unsigned int TNumNodes >
-    static inline void CalculateNuElementMatrix(BoundedMatrix<double, (TDim+1), TNumNodes*(TDim+1)>& rNut,
```
Removes some duplication again! Just for my understanding, this was possible because the offset became equal to TDim here in our new ordering right?

avdg81 replied:
Yes, that is correct. Due to the reordering of the DOFs the offsets became identical for this function and `CalculateNuMatrix`.

WPK4FEM reviewed (commented):
Good step towards uniformly programmed continuum UPw elements.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/element_utilities.hpp:
```diff
@@ -238,9 +213,8 @@ typedef std::size_t IndexType;
 
-//----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-    static inline void AssembleDensityMatrix(Matrix &DensityMatrix,
-                                             double Density)
```
Is there a ublas function that just sets all values of a matrix to a double? I'm not sure if there is a more direct method, but this might work:
```
DensityMatrix = ZeroMatrix(DensityMatrix.size1(), DensityMatrix.size2());
DensityMatrix += Density;
```
Only if you have time left (and if you think it's better), it's not an urgent matter and not sure if it's within scope for this PR.

avdg81 replied:
I agree with you that I would expect there are better ways of doing this. However, I believe this is beyond the scope of the PR. I will add a separate issue to the product backlog such that we can pick it up in the (near?) future.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/element_utilities.hpp:
```diff
@@ -697,6 +594,27 @@ typedef std::size_t IndexType;
+                                    std::size_t        RowOffset,
+                                    std::size_t        ColumnOffset)
+    {
+        for (auto i = std::size_t{0}; i < rSourceMatrix.size1(); ++i) {
+            for (auto j = std::size_t{0}; j < rSourceMatrix.size2(); ++j) {
+                rDestinationMatrix(i + RowOffset, j + ColumnOffset) += rSourceMatrix(i, j);
+            }
+        }
```
Should we do a check somewhere such that we're sure that we don't go out of bounds of the destination matrix? (e.g. source is 3x3, destination 4x4 and offsets are 2)? I'm not sure what happens when using transform and such a thing happens (in the cases of the vector addition)

avdg81 replied:
I believe that such a check is not really needed here. Also the STL algorithms in general don't carry out such checks. They assume it is the caller's responsibility to ensure that the operation won't attempt to access data that is out of bounds. If the programmer forgets about that, you enter the domain of undefined behavior, which may or may not result in a segfault. I would keep it the way it is now.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/interface_element_utilities.hpp:
```diff
@@ -93,49 +93,6 @@ class InterfaceElementUtilities
-                                                const unsigned int& GPoint)
-    {
-        //Quadrilateral_interface_2d_4
-        rNut(0,0) = -Ncontainer(GPoint,0); rNut(0,3) = -Ncontainer(GPoint,1);
```
Just for my understanding, what is the reason for the minus sign compared to the 'normal element' version?

avdg81 replied:
I actually don't know. Perhaps @WPK4FEM has an idea ...?

WPK4FEM replied:
At first glance minus on one side of the interface, plus on the other. Interfaces are about differential movement of the 2 sides.

avdg81 committed:
Processed review feedback by Wijtze Pieter

- Eliminated two temporary variables.
- Renamed several matrices to follow a consistent naming scheme (i.e.
  `UUMatrix`, `UPMatrix`, `PUMatrix`, and `PPMatrix`).
- Renamed two template type names (to `DisplacementNodeRange` and
  `WaterPressureNodeRange`) as well as the corresponding function
  parameters (to `rDisplacementNodes` and `rWaterPressureNodes`).
- The U-Pw diff order elements now reuse the utility function for
  assembling the UP submatrix. To achieve that, the interface of the
  utility function had to be changed. Similar changes have been made to
  the utility functions that assemble the PU and PP submatrices.
- The U-Pw diff order elements now reuse the utility function for
  assembling the U subvector. To achieve that, the interface of the
  utility function had to be changed. A similar change has been made to
  the utility function that assembles the P subvector.
- Renamed two local variables.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_dof_utilities.cpp:
```diff
@@ -169,6 +193,110 @@ KRATOS_TEST_CASE_IN_SUITE(VariableTypeAndNodeIDsMustMatchWhenExtractingDofsFromN
     KRATOS_EXPECT_EQ(dofs[2]->GetId(), 3);
 }
 
+KRATOS_TEST_CASE_IN_SUITE(UDofsPrecedePwDofsWhenExtractingUPwDofsFromNondiffOrder2DElement, KratosGeoMechanicsFastSuite)
```
Very clear new tests!

rfaasse reviewed (commented):
Thank you for this massive clean-up and effort towards more and more consistency! I have some suggestions and questions for my own understanding, none of them blocking (and some might be out of scope as well, so I leave that up to you).

avdg81 committed:
Processed review feedback by Richard

- Replaced a raw loop by an STL algorithm.

rfaasse reviewed (approved):
Looks good to me! Thank you for all the work! There are two open questions left, but they're not blocking in my opinion (still @WPK4FEM could you check if you agree no action is needed for them?).

WPK4FEM reviewed (approved):
Complete enough for me.
